src/cmap/conn/pooled.rs

```diff
@@ -0,0 +1,355 @@
+use std::{
+    ops::{Deref, DerefMut},
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use derivative::Derivative;
+use tokio::sync::{mpsc, Mutex};
+
+use super::{
+    CmapEventEmitter,
+    Connection,
+    ConnectionGeneration,
+    ConnectionInfo,
+    PendingConnection,
+    PinnedConnectionHandle,
+    PoolManager,
+};
+use crate::{
+    bson::oid::ObjectId,
+    cmap::PoolGeneration,
+    error::{Error, Result},
+    event::cmap::{
+        ConnectionCheckedInEvent,
+        ConnectionCheckedOutEvent,
+        ConnectionClosedEvent,
+        ConnectionClosedReason,
+        ConnectionReadyEvent,
+    },
+    runtime::AsyncStream,
+};
+
+/// A wrapper around the [`Connection`] type that represents a connection within a connection pool.
+/// This type derefs into [`Connection`], so fields and methods exposed on that type can be called
+/// directly from this one.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub(crate) struct PooledConnection {
+    /// The connection this pooled connection wraps.
+    connection: Connection,
+
+    /// The connection pool generation from which this connection was checked out.
+    pub(crate) generation: ConnectionGeneration,
+
+    /// Emitter for events related to this connection.
+    #[derivative(Debug = "ignore")]
+    event_emitter: CmapEventEmitter,
+
+    /// The state of this connection.
+    state: PooledConnectionState,
+}
+
+/// The state of a pooled connection.
+#[derive(Clone, Debug)]
+enum PooledConnectionState {
```

Comment:
My original goal for this PR was to make `PooledConnection` generic over its state so that we could assert at compile time that a connection is in the correct state; however, that proved to be a pain because drop specialization is not allowed, and a lot of our connection transformation is done by mutation. This was a compromise solution that required much less overall refactoring, but I still think the additional type information is useful.

Reply:
That's a real shame about drop specialization.  This is still _way_ more readable than before, though.